packages/graphback/src/layers/data/GraphbackDataProvider.ts

```diff
@@ -0,0 +1,35 @@
+
+// If we come with Union on client we might use some complex JSON for describing rules
+// and single key for type for simple use cases
+// tslint:disable-next-line: no-any
+export type AdvancedFilter = any;
+
+
+// TODO pagination
+interface PagedResult<T> {
+    result: T;
+    total: number;
+    page: number;
+}
+
+/**
+ * Builds CRUD interface for data access 
+ */
+// tslint:disable-next-line: no-any
+export interface GraphbackDataProvider<Type = any, GraphbackContext = any> {
+    // FIXME Pagination support
+    // CRUD
+    // FIXME rename to Object = object?
+    createObject(name: string, data: Type, context?: GraphbackContext): Promise<Type>;
+    readObject(name: string, id: string, context?: GraphbackContext): Promise<Type>;
+    updateObject(name: string, id: string, data: Type, context?: GraphbackContext): Promise<Type>;
+    deleteObject(name: string, id: string, context?: GraphbackContext): Promise<string>;
+
+    // Find
+    findAll(name: string, context?: GraphbackContext): Promise<Type[]>;
+    findBy(name: string, filter: Type | AdvancedFilter, context?: GraphbackContext): Promise<Type[]>;
+
+    // Dataloader support
+    // Caching support for nested queries
+    // readObjectWithCache(id: String, data: Type);
+}
```

Comment:
The DataProvider should handle the relationships in order to take advantage of joins

Reply:
Generally there are two ideas that we can use:
- use execution plan of graphql for relationships and make individual queries to db with cache for relationships. 

context.cache['user'][5] can contain cached data

- use root resolvers only and something like our own graphql-query-mapper.
This approach would work nicely with joins, but it will require devs to build their own equivalent of execution plan. 

For the moment we going to take approach nr1. I think you have really interesting ideas. If you have some article or blog that gives.more.info let me know

-